include/kernels/madwf_tensor.cuh

```diff
@@ -0,0 +1,205 @@
+      const int Ls_in;  // length of 5th dimension of the in field
+
+      const int volume_4d_cb;
+
+      matrix_type *reduce_buffer = nullptr;
+
+      const int nParity;
+
+      matrix_type *result_d;
+
+      int batch;
+
+      Tensor5DArg(const ColorSpinorField &out, const ColorSpinorField &in, int num_x_blocks, matrix_type *result_d) :
+        kernel_param(dim3(out.VolumeCB() / out.X(4), out.X(4), out.SiteSubset())),
+        out(out),
+        in(in),
+        Ls_out(out.X(4)),
+        Ls_in(in.X(4)),
+        volume_4d_cb(in.VolumeCB() / in.X(4)),
+        nParity(in.SiteSubset()),
+        result_d(result_d),
+        batch(num_x_blocks)
+      {
+
+        if (volume_4d_cb != static_cast<int>(out.VolumeCB() / Ls_out)) {
+          errorQuda("Input and Output fields should have the same 4d volume: %d neq %d.\n", volume_4d_cb,
+                    static_cast<int>(out.VolumeCB() / Ls_out));
+        }
+
+        if (in.Nspin() != 4) errorQuda("nSpin = %d not support", in.Nspin());
+        if (in.Ncolor() != 3) errorQuda("nColor = %d not support", in.Ncolor());
+        if (out.Nspin() != 4) errorQuda("nSpin = %d not support", out.Nspin());
+        if (out.Ncolor() != 3) errorQuda("nColor = %d not support", out.Ncolor());
+
+        if (!in.isNative() || !out.isNative())
+          errorQuda("Unsupported field order out=%d in=%d\n", out.FieldOrder(), in.FieldOrder());
+
+        size_t reduce_bytes = num_x_blocks * sizeof(matrix_type) * out.X(4) * in.X(4);
+        reduce_buffer = reinterpret_cast<matrix_type *>(device_malloc(reduce_bytes));
+      }
+
+      ~Tensor5DArg()
+      {
+        if (reduce_buffer) { device_free(reduce_buffer); }
+      }
+    };
+
+    template <class Arg> struct Tensor5D {
+
+      const Arg &arg;
+      constexpr Tensor5D(const Arg &arg) : arg(arg) { }
+      static constexpr const char *filename() { return KERNEL_FILE; }
+
+      /**
+        @brief Form a Ls_in-by-Ls_out tensor product from the two input vectors. The type of reduce_buffer
+                determines if the resulting tensor has spin and/or color d.o.f, or just two chiral d.o.f
+        @param[in] parity Parity we are on
+        @param[in] x_b Checkerboarded 4-d space-time index
+        @param[in] s Ls dimension coordinate
+       */
+      __device__ __host__ inline void operator()(int x_cb, int s, int parity)
+      {
+        using Vector = typename Arg::Vector;
+
+        const int Ls_in = arg.Ls_in;
+        const int Ls_out = arg.Ls_out;
+        const int volume_4d_cb = arg.volume_4d_cb;
+        auto reduce_buffer = arg.reduce_buffer;
+
+        if (x_cb >= volume_4d_cb) return;
+        if (s >= Ls_out) return;
+        if (parity >= arg.nParity) return;
+
+        SharedMemoryCache<Vector> cache;
+
+        int ld = Ls_in * target::block_dim().x;
+        int t = s;
+        while (t < Ls_in) {
+          int index = t * target::block_dim().x + target::thread_idx().x;
```

Comment:
Just wondering here why the standard interface for the cache load/save can't be used here?

Reply:
Since the standard load/save interface is based on the current block shape. Here the `y` dimension is based on the Ls_in instead of `block_dim().y`.

Reply:
The block size that the interface utilizes can be set in the constructor for the `SharedMemoryCache`, so it doesn't have to match the block size.  Or does this not work in this case for some reason?

Reply:
https://github.com/lattice/quda/pull/1061/commits/d776b593adac861010d314b14523ba261ada3c14